File: src/typescript/shared/telegram_auth.ts
import * as crypto from 'crypto';

export const generateOTP = (seed: string, timestamp?: number): string => {
  const ts = timestamp || Math.floor(Date.now() / 1000);
  const message = Math.floor(ts / 30).toString();
  return crypto.createHmac('sha256', seed).update(message).digest('hex').substring(0, 8);
};

export const verifyOTP = (seed: string, otp: string, window: number = 1): boolean => {
  const currentTs = Math.floor(Date.now() / 1000);
  const currentWindow = Math.floor(currentTs / 30);
  for (let delta = -window; delta <= window; delta++) {
    const testWindow = currentWindow + delta;
    const testOtp = crypto.createHmac('sha256', seed).update(testWindow.toString()).digest('hex').substring(0, 8);
    if (crypto.timingSafeEqual(Buffer.from(testOtp, 'hex'), Buffer.from(otp, 'hex'))) {
      return true;
    }
  }
  return false;
};

File: src/python/shared/telegram_auth.py
import hmac
import hashlib
import time
from typing import Optional


def generate_otp(seed: str, timestamp: Optional[int] = None) -> str:
    """Generate HMAC-signed OTP per Section 4.3"""
    ts = timestamp or int(time.time())
    # 30-second window
    message = str(ts // 30).encode()
    return hmac.new(seed.encode(), message, hashlib.sha256).hexdigest()[:8]


def verify_otp(seed: str, otp: str, window: int = 1) -> bool:
    """Verify OTP with time window tolerance"""
    current_ts = int(time.time())
    current_window = current_ts // 30
    for delta in range(-window, window + 1):
        test_window = current_window + delta
        test_otp = hmac.new(
            seed.encode(), str(test_window).encode(), hashlib.sha256
        ).hexdigest()[:8]
        if hmac.compare_digest(test_otp, otp):
            return True
    return False
